### PR TITLE
Refactor: Remove Named Return Variables

### PR DESCRIPTION
**Pull request by sarthaksarthak9**
Remove Named Returns for Better Code Maintainability
#11446

### CODE REVIEW

File: pkg/instancetype/instancetype.go [hunk 1]
```diff
@@ -811,39 +811,41 @@ Volume -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourcePVC -> Persi
...
-func (m *InstancetypeMethods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameLabel, defaultKindLabel string) (defaultName, defaultKind string, err error) {
+func (m *InstancetypeMethods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameLabel, defaultKindLabel string) (string, string, error) {
 	for _, volume := range vm.Spec.Template.Spec.Volumes {
 		if volume.Name != inferFromVolumeName {
 			continue
 		}
 		if volume.PersistentVolumeClaim != nil {
-			return m.inferDefaultsFromPVC(volume.PersistentVolumeClaim.ClaimName, vm.Namespace, defaultNameLabel, defaultKindLabel)
+			defaultName, defaultKind, err := m.inferDefaultsFromPVC(volume.PersistentVolumeClaim.ClaimName, vm.Namespace, defaultNameLabel, defaultKindLabel)
+			return defaultName, defaultKind, err
```
Comment (vasiliy-ul) [inline pkg/instancetype/instancetype.go:821]: Why do we need to change this? Doing just `return m.inferDefaultsFromPVC(...)` should be fine here.

File: pkg/instancetype/instancetype.go [hunk 2]
```diff
@@ -811,39 +811,41 @@ Volume -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourcePVC -> Persi
...
 		}
 		if volume.DataVolume != nil {
-			return m.inferDefaultsFromDataVolume(vm, volume.DataVolume.Name, defaultNameLabel, defaultKindLabel)
+			defaultName, defaultKind, err := m.inferDefaultsFromDataVolume(vm, volume.DataVolume.Name, defaultNameLabel, defaultKindLabel)
+			return defaultName, defaultKind, err
```
Comment (vasiliy-ul) [inline pkg/instancetype/instancetype.go:825]: Same as above.

File: pkg/instancetype/instancetype.go [hunk 3]
```diff
@@ -855,21 +857,19 @@ func (m *InstancetypeMethods) inferDefaultsFromDataVolume(vm *virtv1.VirtualMach
 	}
 	dv, err := m.Clientset.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Get(context.Background(), dvName, metav1.GetOptions{})
 	if err != nil {
-		// Handle garbage collected DataVolumes by attempting to lookup the PVC using the name of the DataVolume in the VM namespace
```
Comment (vasiliy-ul) [inline pkg/instancetype/instancetype.go:858]: Why do we need to drop the comments? There are several more similar changes further on.

**Comment by iholder101** [comment]
Hi @sarthaksarthak9! Thank you very much for your contribution!

Please see my comment here: https://github.com/kubevirt/kubevirt/issues/11446#issuecomment-1980743664. You're more than welcome to express your opinion on the issue.

**Comment by iholder101** [comment]
Let's have a general discussion [here](https://github.com/kubevirt/kubevirt/issues/11446#issuecomment-1980743664), as I don't think the implementation itself is problematic, but the overall notion that returned parameters are necessarily bad.

/hold

**Comment by alicefr** [comment]
@iholder101 we should discuss it in the coding best practise issue that @0xFelix opened: https://github.com/kubevirt/kubevirt/issues/11259 . However, I got multiple times the feedback that we should avoid this, e.g https://github.com/kubevirt/kubevirt/pull/11260#discussion_r1494465615. We have just few places in the code where we still have this pattern and we should try to uniform this. 
I also agree with the other that they are a bit confusing.

File: pkg/instancetype/instancetype.go (hunk 3 above)
Comment (sarthaksarthak9) [reply pkg/instancetype/instancetype.go:858]: @vasiliy-ul I by mistake drop this comments

File: pkg/instancetype/instancetype.go (hunk 1 above)
Comment (sarthaksarthak9) [reply pkg/instancetype/instancetype.go:821]: I thought it improves readability of the code

File: pkg/instancetype/instancetype.go [hunk 4]
```diff
@@ -811,39 +811,41 @@ Volume -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourcePVC -> Persi
 Volume -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourceRef -> DataSource
 Volume -> DataVolumeSource -> DataVolumeTemplate -> DataVolumeSourceRef -> DataSource -> PersistentVolumeClaim
 */
-func (m *InstancetypeMethods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameLabel, defaultKindLabel string) (defaultName, defaultKind string, err error) {
+func (m *InstancetypeMethods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameLabel, defaultKindLabel string) (string, string, error) {
```
Comment (EdDev) [inline pkg/instancetype/instancetype.go:814]: The linter will not allow two or more returned parameters from the same type to be unnamed.

**Review by EdDev — CHANGES_REQUESTED** [review]
Unfortunately, this attempt to remove named parameters will not work.

**Comment by EdDev** [comment]
> Let's have a general discussion [here](https://github.com/kubevirt/kubevirt/issues/11446#issuecomment-1980743664), as I don't think the implementation itself is problematic, but the overall notion that returned parameters are necessarily bad.
> 
> /hold

I do not think this merits a hold, it is either an ack or nack.

Like many other language features, there are pros and cons to everything. Unfortunately, many features are missuses or overused.
The context of everything matters and in this specific case the existance of the named parameters has value.

I would argue that the reason it is needed is the real problem, i.e. returning more than one value (not including the error), but that also depends on a case per case scenario.

**Comment by thebigbone** [comment]
Will the use of [variadic function](https://gobyexample.com/variadic-functions) make sense here?

**Comment by EdDev** [comment]
> Will the use of [variadic function](https://gobyexample.com/variadic-functions) make sense here?

I do not understand.
The subject is about returned values, not input args.

**Comment by thebigbone** [comment]
> > Will the use of [variadic function](https://gobyexample.com/variadic-functions) make sense here?
> 
> I do not understand. The subject is about returned values, not input args.

You're right. I misunderstood the topic.

**Pull request closed without merging**